File: src/Tests/Performance/PerfMatMul/PerfMatMul.go
package main

import (
	"fmt"
	"time"
)

// Dense float64 matrix multiply: measures floating-point throughput and
// nested slice-of-slice element access.

func matmul(n int) float64 {
	a := make([][]float64, n)
	b := make([][]float64, n)
	c := make([][]float64, n)

	for i := 0; i < n; i++ {
		a[i] = make([]float64, n)
		b[i] = make([]float64, n)
		c[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			a[i][j] = float64((i*n+j)%100) * 0.5
			b[i][j] = float64((i+j)%100) * 0.25
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			sum := 0.0
			for k := 0; k < n; k++ {
				sum += a[i][k] * b[k][j]
			}
			c[i][j] = sum
		}
	}

	trace := 0.0
	for i := 0; i < n; i++ {
		trace += c[i][i]
	}
	return trace
}

func main() {
	start := time.Now().UnixNano()

	total := 0.0
	for r := 0; r < 4; r++ {
		total += matmul(256)
	}

	elapsed := time.Now().UnixNano() - start
	fmt.Println("checksum:", int64(total))
	fmt.Println("elapsed_ns:", elapsed)
}

File: src/Tests/Performance/PerfMatMul/PerfMatMul.cs
namespace go;

using fmt = fmt_package;
using time = time_package;

partial class main_package {

internal static float64 matmul(nint n) {
    var a = new slice<slice<float64>>(n);
    var b = new slice<slice<float64>>(n);
    var c = new slice<slice<float64>>(n);
    for (nint i = 0; i < n; i++) {
        a[i] = new slice<float64>(n);
        b[i] = new slice<float64>(n);
        c[i] = new slice<float64>(n);
        for (nint j = 0; j < n; j++) {
            a[i][j] = (float64)((i * n + j) % 100) * 0.5D;
            b[i][j] = (float64)((i + j) % 100) * 0.25D;
        }
    }
    for (nint i = 0; i < n; i++) {
        for (nint j = 0; j < n; j++) {
            var sum = 0.0D;
            for (nint k = 0; k < n; k++) {
                sum += a[i][k] * b[k][j];
            }
            c[i][j] = sum;
        }
    }
    var trace = 0.0D;
    for (nint i = 0; i < n; i++) {
        trace += c[i][i];
    }
    return trace;
}

internal static void Main() {
    var start = time.Now().UnixNano();
    var total = 0.0D;
    for (nint r = 0; r < 4; r++) {
        total += matmul(256);
    }
    var elapsed = time.Now().UnixNano() - start;
    fmt.Println("checksum:", (int64)total);
    fmt.Println("elapsed_ns:", elapsed);
}

} // end main_package
